Improve plotting: separate subplots per benchmark + add gate ratio metric

AdiBak opened the pull request:
## Summary
Improves the plotting system by separating benchmarks into individual subplots for better readability, as requested in #156.

## Changes Made
- Changed layout from single-axis plots to clean 2x3 grid with individual subplots per benchmark
- New compilation metric showing ratio `compiled_gates / raw_gates` 
- Generates individual files for compile-time, gate-count, gate-ratio, rel-err-ideal, rel-err-noisy
- Added option to use linear vs log scale (gate ratio uses linear, others use log)

## Files Changed
- `plotting/plot_latest_benchmark.py` - Updated plotting functions to use subplot layout and add gate ratio metric

## Technical Details
- Replaced `generate_plot()` with `generate_compilation_subplots()` and `generate_simulation_subplots()`
- Changed from dynamic row calculation to fixed 2x3 grid layout
- Added `use_log_scale` configuration option to plot configs
- Added compiled gate ratio calculation: `compiled_multiq_gates / raw_multiq_gates`

## Key Improvements
- Each benchmark has its own subplot  
- 2x3 layout for easy comparison  
- Linear scale for gate ratio (values typically 0-1.5) 
- Separate files per metric for organization  

## Images
### Compilation:
<img width="4469" height="2366" alt="latest_compiler_benchmarks_by_circuit_compiled-ratio" src="https://github.com/user-attachments/assets/9eb007ce-09b6-4838-8e96-03e7d4b48084" />
<img width="4468" height="2366" alt="latest_compiler_benchmarks_by_circuit_compiled-multiq-gates" src="https://github.com/user-attachments/assets/9d4414b4-174c-4062-972e-bec3cd791bc3" />
<img width="4468" height="2366" alt="latest_compiler_benchmarks_by_circuit_compile-time" src="https://github.com/user-attachments/assets/4a03b79b-f304-49f7-8ab7-ba3f0fc6caa0" />

### Simulation:
<img width="4469" height="2366" alt="latest_simulation_benchmarks_by_circuit_rel-err-noisy" src="https://github.com/user-attachments/assets/2e8e20c4-2410-42f3-b2b8-6ababdeb386a" />
<img width="4469" height="2366" alt="latest_simulation_benchmarks_by_circuit_rel-err-ideal" src="https://github.com/user-attachments/assets/878f3d6e-9548-47d2-a0e8-e39255d78b64" />

## Testing
Tested with latest benchmark data from `ucc-benchmarks-8-core-U22.04` runner. 

## Issue Addressed
#156.

---
*Thanks so much to @jordandsullivan for the guidance and feedback!*

AdiBak commented:
Hi,
Thanks for the feedback! I refactored the code to have a helper plotting function `generate_subplots` that is now used for compilation and simulation subplot generation, and removed the unused `bars` variable.

bachase commented on plotting/plot_latest_benchmark.py:
```diff
@@ -27,62 +28,112 @@ def generate_plot(
+    use_pdf: bool = False,
+):
+    """Generate subplots for compilation benchmarks with separate subplot per benchmark."""
+    generate_subplots(df, plot_configs, latest_date, out_path, use_pdf)
```
Since `generate_compilation_subplots` (and `generate_compilation_subplots` later on) just directly call `generate_subplots`, I would just eliminate these functions and gall `generate_subplots` directly.

bachase commented on plotting/plot_latest_benchmark.py:
```diff
@@ -27,62 +28,112 @@ def generate_plot(
+                if config.get("use_log_scale", True):
+                    ax.set_yscale("log")
+            else:
+                ax.set_visible(False)
```
If there are more than 6 benchmark results, does this silently just not show the additional plots? I'd consider either supporting arbitrary number of benchmark results, or at least asserting/erroring its not the hard coded 6 results.

AdiBak replied:
Sure! I made it dynamically determine how many rows and cols should exist, so essentially it'd follow this:
- 1-3 benchmarks: 1 row, 1-3 columns
- 4-6 benchmarks: 2 rows, 3 columns (same as before)
- 7-9 benchmarks: 3 rows, 3 columns
- 10-12 benchmarks: 3 rows, 4 columns
- More than 12: Calculate optimal grid

Let me know what you think. Thanks!

bachase commented on plotting/plot_latest_benchmark.py:
```diff
@@ -27,62 +28,112 @@ def generate_plot(
             "y_col": "compiled_multiq_gates",
             "title": "Gate Counts",
             "ylabel": "Compiled Gate Count",
+            "use_log_scale": True,
```
It might be better to have `ylabel` be "Compiled Multi-Qubit Gate Count" to be clear it doesn't include single qubit gates.

bachase reviewed (commented):
Thanks for this work @AdiBak. I left a few comments/questions.

For overall feedback, for the compiler metrics (gate count/compile-time), I don't find the format more convenient to read than we have now for a quick overview/glance of performance. I think it does show the "Gate Counts" plot isn't so good, since it mixes scales of different circuits. I would consider replacing that with a compiled ratio plot instead, but keeping the same "all in one plot" format.

For the simulation plots, I also like the style in #156 for the compiled vs. uncompiled vs ideal, but not that isn't quite what this PR does. Regardless, I wouldn't include the simulation ones in the README.md until we better calibrate the noise levels. 

If we want to switch to this style, I'd lean towards making a `BENCHMARK.md` or something which links to all the images, adds a little context on how to interpret them, and link to that from the README and the docs.

All that being said, I don't mind auto-generating these in addition to what we have today, but I'm not in favor of just replacing directly.

jordandsullivan commented:
@AdiBak let us know if you need any help with the changes @bachase suggested. I know it's your first time contributing, so we want to make sure you have the support you need :)

AdiBak commented:
Hi @jordandsullivan! Yes, my apologies for the delay. Thanks for checking in!

I implemented some of the changes:
- Replaced gate counts with compiled ratio in the all-in-one format (keeping the same grouped bar chart style)
- Added raw gate counts to x-axis labels - each benchmark now shows the name and original 2-qubit gate count on a second line
- Generated subplot versions as additional plots which are saved with `_subplots` suffix

The all-in-one type plots of `compile-time` and `compiled-ratio` (replacing the gate counts one as Brad suggested) should show now, and all three metrics (compile-time, gate-counts, compiled-ratio) are still there as additional plots. 

Here are the new all-in-one plots showing the updated format:

<img width="3568" height="1770" alt="latest_compiler_benchmarks_by_circuit_compiled-ratio" src="https://github.com/user-attachments/assets/91fde441-77dc-4844-bda1-a2f099e4f247" />
<img width="3566" height="1770" alt="latest_compiler_benchmarks_by_circuit_compile-time" src="https://github.com/user-attachments/assets/e127793d-3efe-4fac-b06b-675925836b1f" />

Please let me know what you think. Thanks again!

AdiBak commented:
Thanks!

Certainly, here are the simulation plots:

<img width="3566" height="1770" alt="latest_simulation_benchmarks_by_circuit_rel-err-noisy" src="https://github.com/user-attachments/assets/925e0c45-83be-4c71-a8e1-d57b6b8aaac0" />
<img width="3565" height="1770" alt="latest_simulation_benchmarks_by_circuit_rel-err-ideal" src="https://github.com/user-attachments/assets/3af985c7-26ac-41e5-8335-e1bb52b99889" />
